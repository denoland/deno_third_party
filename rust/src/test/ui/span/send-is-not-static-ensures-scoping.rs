// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

struct Guard<'a> {
    f: Box<Fn() + Send + 'a>,
}

fn scoped<'a, F: Fn() + Send + 'a>(f: F) -> Guard<'a> {
    Guard { f: Box::new(f) }
}

impl<'a> Guard<'a> {
    fn join(self) {}
}

fn main() {
    let bad = {
        let x = 1;
        let y = &x;
        //~^ ERROR `x` does not live long enough

        scoped(|| {
            let _z = y;
            //~^ ERROR `y` does not live long enough
        })
    };

    bad.join();
}
