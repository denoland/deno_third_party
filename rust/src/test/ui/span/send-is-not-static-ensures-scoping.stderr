error[E0597]: `x` does not live long enough
  --> $DIR/send-is-not-static-ensures-scoping.rs:16:18
   |
LL |         let y = &x;
   |                  ^ borrowed value does not live long enough
...
LL |     };
   |     - `x` dropped here while still borrowed
...
LL | }
   | - borrowed value needs to live until here

error[E0597]: `y` does not live long enough
  --> $DIR/send-is-not-static-ensures-scoping.rs:20:22
   |
LL |         scoped(|| {
   |                -- capture occurs here
LL |             let _z = y;
   |                      ^ borrowed value does not live long enough
...
LL |     };
   |     - borrowed value only lives until here
...
LL | }
   | - borrowed value needs to live until here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
